text/3519-arbitrary-self-types-v2.md

```diff
@@ -297,7 +358,44 @@ Rust standard library smart pointers are designed with this shadowing behavior i
 
 Furthermore, the `Deref` trait itself [documents this possible compatibility hazard](https://doc.rust-lang.org/nightly/std/ops/trait.Deref.html#when-to-implement-deref-or-derefmut), and the Rust API Guidelines has [a guideline about avoiding inherent methods on smart pointers](https://rust-lang.github.io/api-guidelines/predictability.html#smart-pointers-do-not-add-inherent-methods-c-smart-ptr).
 
-These method shadowing risks also apply to `Receiver`. This RFC does not make things worse for types that implement `Deref`, it only adds additional flexibility to the `self` parameter type for `T::m`.
+This RFC does not make things worse for types that implement `Deref`.
+
+_However_, this RFC allow types to implement `Receiver`, and in fact does so for `NonNull<T>` and `Weak<T>`. `NonNull<T>` and `Weak<T>` were not designed with method shadowing concerns in mind. This would run the risk of breakage:
```

Comment:
From my understanding, there isn't yet consensus on allowing raw pointers/`NonNull`/`Weak`/others, so I was thinking that it might be nice to basically include both proposals in the RFC, to give a nice comparison of the tradeoffs between them, something like `arbitrary_self_types_minimal` and `arbitrary_self_types_full`?

As an example, the RFC text could be done as follows:

```suggestion
**Minimal:** These method shadowing risks also apply to `Receiver`. This RFC does not make things worse for types that implement `Deref`, it only adds additional flexibility to the `self` parameter type for `T::m`.

**Full:** This RFC does not make things worse for types that implement `Deref`.

    _However_, this RFC allow types to implement `Receiver`, and in fact does so for `NonNull<T>` and `Weak<T>`. `NonNull<T>` and `Weak<T>` were not designed with method shadowing concerns in mind. This would run the risk of breakage:
```

Along with a section in the top of the RFC explaining the use of `**Minimal**` and `**Full**` (unsure of the exact formatting)?

Reply:
I would like to do this, but I'm told that there in fact might be a consensus here, which is that key members of T-Lang really really want to support it for raw pointers/`NonNull`/`Weak`, and that if we don't manage to succeed in adding that support, we might not get support for arbitrary self types at all. I think that would be a shame so I am poking at this belief in some 1:1s — I'll get back to this comment once I've got a better sense of folks' opinions!